jfix-zookeeper/src/main/kotlin/ru/fix/zookeeper/instance/registry/ServiceInstanceIdRegistry.kt

```diff
@@ -0,0 +1,161 @@
+ * Service registration guarantees uniqueness of instance id in case of parallel startup on different JVMs.
+ */
+class ServiceInstanceIdRegistry(
+        private val curatorFramework: CuratorFramework,
+        private val instanceIdGenerator: InstanceIdGenerator,
+        private val config: ServiceInstanceIdRegistryConfig,
+        profiler: Profiler
+) : AutoCloseable {
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(ServiceInstanceIdRegistry::class.java)
+    }
+
+    private val registeredInstanceIdLocks: MutableSet<LockIdentity> = ConcurrentSkipListSet()
+
+    private val lockManager: PersistentExpiringLockManager = PersistentExpiringLockManager(
+            curatorFramework,
+            DynamicProperty.of(
+                    PersistentExpiringLockManagerConfig(
+                            lockAcquirePeriod = config.disconnectTimeout.dividedBy(2),
+                            expirationPeriod = config.disconnectTimeout.dividedBy(3),
+                            lockCheckAndProlongInterval = config.disconnectTimeout.dividedBy(4),
+                            acquiringTimeout = Duration.ofSeconds(1)
+                    )
+            ),
+            profiler
+    )
+
+    init {
+        curatorFramework
+                .connectionStateListenable
+                .addListener(
+                        ConnectionStateListener { _, newState ->
+                            when (newState) {
+                                ConnectionState.RECONNECTED -> reconnect()
+                                else -> Unit
+                            }
+                        })
+        initServiceRegistrationPath()
+    }
+
+    /**
+     * @param serviceName name of service should be registered
+     * @return generated instance id for registered service
+     */
+    fun register(serviceName: String): String {
+        for (i in 1..config.countRegistrationAttempts) {
+            val alreadyRegisteredInstanceIds = curatorFramework.children.forPath(config.serviceRegistrationPath)
+            val preparedInstanceId = instanceIdGenerator.nextId(alreadyRegisteredInstanceIds)
+
+            val instanceIdPath = ZKPaths.makePath(config.serviceRegistrationPath, preparedInstanceId)
+            val instanceIdData = InstanceIdData(serviceName, Instant.now())
+            val lockIdentity = LockIdentity(instanceIdPath, Marshaller.marshall(instanceIdData))
+
+            val result = lockManager.tryAcquire(lockIdentity) { lock ->
+                logger.warn("Failed to prolong lock=$lock for service=$serviceName")
+            }
+            if (result) {
+                registeredInstanceIdLocks.add(lockIdentity)
+                logger.info("Instance of service=$serviceName started with id=$preparedInstanceId")
+                return preparedInstanceId
+            } else if (i == config.countRegistrationAttempts) {
+                logger.error("Failed to register service=$serviceName. " +
```

Comment:
It is quite common to start two application instance almost simultaneously. Suggestion: for first retry make it debug. For retry attempt 2, 3.. make it warning and for last failed one make it error.

Reply:
fixed